applybn/feature_selection/bn_nmi_feature_selector.py

```diff
@@ -0,0 +1,166 @@
+    Hab = self.entropy_map[(a.name, b.name)]
+    return (Ha + Hb - Hab) / (2 * max(Ha, Hb))
+
+  def _prepare_dataset(self, df: pd.DataFrame):
+    '''
+    Prepare the dataset, discretize the fields, remove rows
+    Return a ready-to-use dataset
+    
+    Args: 
+        df - data frame to be cleaned.
+    '''
+    clean_df = pd.DataFrame()
+    for col in df:
+      if pd.api.types.is_integer_dtype(df[col].dtype):
+        # integers can be discreticised this way, without binning
+        df_map = { l:i for i,l in enumerate(df[col].unique())}
+        clean_df[col] = df[col].map(df_map)
+      elif pd.api.types.is_numeric_dtype(df[col].dtype):
+        # discretisize to bin_count bins
+        clean_df[col] = pd.cut(df[col], bins=self.bin_count, labels=self.bin_labels, duplicates='drop')
+        clean_df[col] = clean_df[col].cat.add_categories(self.bin_count + 1).fillna(self.bin_count + 1)
+      else:
+        # pandas cut does not work with string features, this is done manually
+        # it is assumed that the user has removed the ID-like features, otherwise there will be a memory error later
+        df_map = { l:i for i,l in enumerate(df[col].unique())}
+        clean_df[col] = df[col].map(df_map)
+    return clean_df
+
+  def fit(self, X: pd.DataFrame, y: pd.Series = None):    
+    if y is None:
+      return self
+
+    df = X
+    df[y.name] = y
+    target = y.name
+    self.initial_features = list(df.columns)
+
+    clean_df = self._prepare_dataset(df)
+    if self.verboose:
+      logging.info('Dataset is ready')
+    nmi_map = dict()
+    self.entropy_map = dict()
+    pre_features = []
+    for col in clean_df:
+      if col == target:
+        continue
+      if (col, target) not in nmi_map.keys():
+        nmi_map[(col, target)] = self._normalized_mutual_information(clean_df[col], clean_df[target])
+      nmi = nmi_map[(col, target)]
+      if nmi > self.threshhold:
+        pre_features.append(col)
+
+    if self.verboose:
+      logging.info(f'First selection done, {len(pre_features)} selected', pre_features)
+
+    if len(pre_features) == 1:
+      self.features_selected = pre_features
+      self.feature_names_in_ = X.columns
+      self.n_features_in_ = X.shape[1]
+      self.is_fitted_ = True
+      self.X_ = X
+      self.y_ = y
+      return self
+
+    features = []
+    for idx1, col1 in enumerate(pre_features):
+      for idx2 in range(idx1, len(pre_features)):
+        col2 = pre_features[idx2]
+        if col1 == col2 or col1 == target or col2 == target:
+          continue
+
+        if (col1, col2) not in nmi_map:
+          nmi_map[(col1, col2)] = self._normalized_mutual_information(clean_df[col1], clean_df[col2])
+
+        nmi_tg_1 = nmi_map[(col1, target)]
+        nmi_tg_2 = nmi_map[(col2, target)]
+        nmi_col_12 = nmi_map[(col1, col2)]
+
+        if (nmi_tg_1 > nmi_tg_2) and (nmi_col_12 > nmi_tg_2):
+          if col1 not in features:
+            features.append(col1)
```

Comment:
nested loops here also. Consider refactoring that with pandas methods to avoid nested loops.